src/doc/imageoutput.tex

```diff
@@ -1167,6 +1167,96 @@ \subsection{Copying an entire image}
 \end{code}
 
 
+\subsection{Writing the file to a memory buffer}
+\label{sec:imageoutput:writefiletomemory}
+\index{writing an image file to memory buffer}
+\NEW   % 1.9
+
+Some file format writers allow you to write the output to an in-memory
+buffer rather than a stored file on disk. The buffer is not merely the pixel
+array, but is a byte-for-byte representation of the correctly formatted file
+that would have been written.
+
+This is very incompletely implemented -- currently, only PNG supports this
+feature, though others may be supported in the future (based on demand
+and volunteers).
+
+To find out if a particular file format supports this feature, you can create
+an \ImageOutput of the right type, and check if it supports the feature
+name \qkw{write_memory}:
+
+\begin{code}
+    ImageOutput *out = ImageOutput::create (filename);
+    if (! out  ||  ! out->supports ("write_memory")) {
+        ImageOutput::destroy (out);
+        out = nullptr;
+        return;
+    }
+\end{code}
+
+\ImageOutput writers that support \qkw{write_memory} will respond to two
+special attributes: (1) \qkw{oiio:write_memory}, which passes a pointer to a
+{\cf unsigned char*} that will be modified to contain the address of the
+allocated memory block that will contain the file image bytes; (2)
+\qkw{oiio:write_memory_size}, which passes a pointer to a {\cf uint64_t}
+that will hold the size (in bytes) of the block.
+
+You just add these two attributes to the \ImageSpec that you pass when you
+open the file. Note that you aren't passing the \emph{values} of your
+variables, you are passing their \emph{addresses}, and this extra level of
+indirection requires an unusual idiom in setting up the attributes:
+
+\begin{code}
+    // ImageSpec describing the image we want to write.
+    ImageSpec spec (xres, yres, channels, TypeDesc::UINT8);
+
+    unsigned char *file_buffer = nullptr;  // will get ptr to the memory
+    uint64_t file_buffer_size = 0;         // will get the size
+    void *ptr = &file_buffer;
+    spec.attribute ("oiio:write_memory", TypeDesc::PTR, &ptr);
```

Comment:
It's subtle. OK, so the general pattern is

    attribute (name, type, ptr_to_data_of_type)

So to pass a float or a float array, you would do

    float f;
    float fa[4];
    attribute ("f", TypeDesc(FLOAT), &f);
    attribute ("fa", TypeDesc(FLOAT,4), &fa);   // or just `fa` -- in C/C++, an array name is its address

Note a subtlety: you pass a scalar by address, not by value, so that the call sequence is the same if it was an array you are passing.

What the data you want to pass is itself a pointer?

    char *buffer = 0x10001000;    // buffer points to address 0x10001000
    attribute ("ptr", TypeDesc::PTR, &buffer);

That's fine if you want OIIO to know the address of an existing buffer (0x10001000), or of course it could also read data from the buffer itself.

But in this case, it's OIIO that is allocating the buffer memory, we need a way to pass back the address of the memory we allocated. We need to WRITE that somewhere. If we pass &buffer as above, we can write where `buffer` pointed to, but we have no way to change the caller's `buffer` variable, if you see what I mean. We need one more level of indirection.

    char *buffer;
    char **bufferptr = &buffer;
    attribute ("ptr", TypeDesc::PTR, &bufferptr);

So now the data we're passing as the attribute is a pointer, but it's not a pointer to the buffer (which doesn't exist yet), it's a pointer to the pointer where we will store the address of the buffer we allocate, so that the calling application can then take ownership of the buffer.

It is admittedly awkward and is a weird corner case for the calling convention, which was not really designed to pass data to OIIO, but not have a way for OIIO to be able to modify the variables owned by the app.